FreeBSD and DragonFly: Indirect follow-up on #186

Hi Vincent

I didn't get the chance to update the FreeBSD ports until now but realized that #186 has somehow come back or actually wasn't gone. I have the strong guess that I didn't thorougly or consistently check with all options back then, my bad. 

Only when I tried all options individually and not the default set of options in the port Makefile that I discovered the build consistently failed only if ``--with-snmp`` was given.

I could confirm this with FreeBSD 10.3, 11.0 and DragonFly BSD 4.8 (on amd64). 

## Minimal steps to reproduce the problem
1. Get the tarball of 0.9.7 (or any version later than 0.9.4 from git, then using autogen.sh)
2. Have net-snmp installed (``pkg install net-snmp``)
3. Compile `lldpd` with `./configure --with-snmp && make`
4. The build fails only if configured ``--with-snmp``, not if not enabled as follows like in #186:

```
interfaces-bsd.c: In function 'ifbsd_macphy':
interfaces-bsd.c:474:2: warning: implicit declaration of function 'typeof' [-Wimplicit-function-declaration]
  typeof(ifmr.ifm_ulist[0]) media_list[32] = {};
  ^
interfaces-bsd.c:474:28: error: expected ';' before 'media_list'
  typeof(ifmr.ifm_ulist[0]) media_list[32] = {};
                            ^
compilation terminated due to -Wfatal-errors.
*** Error code 1
```

The current "working" hack I found to be working on Dragonfly and FreeBSD was as follows - but I'm uncertain if I'm not causing unknown damage elsewhere:
```
--- src/daemon/interfaces-bsd.c.orig    2017-06-11 23:00:17.000000000 +0200
+++ src/daemon/interfaces-bsd.c 2017-06-11 22:56:55.000000000 +0200
@@ -470,7 +470,7 @@
 {
 #ifdef ENABLE_DOT3
        struct ifmediareq ifmr = {};
-#ifdef HAVE_TYPEOF
+#ifdef HOST_OS_OPENBSD
        typeof(ifmr.ifm_ulist[0]) media_list[32] = {};
 #else
        int media_list[32] = {};
```
If HOST_OS_OPENBSD works instead of HAVE_TYPEOF, commit bdc8f796579f0576b62166ed71c761510d9a2499 could be reverted. For regression testing I built lldpd with the exact same configure options on OpenBSD 6.1 where it didn't break the build with or without this change. I did miss NetBSD which I hadn't at hand.

What's your opinion on this?

--- a/src/daemon/Makefile.am
+++ b/src/daemon/Makefile.am
@@ -100,13 +100,6 @@ liblldpd_la_SOURCES += \
 	priv-bsd.c
 endif
 
-# Add SNMP support if needed
-if USE_SNMP
-liblldpd_la_SOURCES += agent.c agent_priv.c agent.h
-liblldpd_la_CFLAGS  += @NETSNMP_CFLAGS@
-lldpd_LDADD         += @NETSNMP_LIBS@
-endif
-
 # seccomp support
 if USE_SECCOMP
 BUILT_SOURCES += syscall-names.h
@@ -121,6 +114,16 @@ nodist_liblldpd_la_SOURCES += syscall-names.h
 liblldpd_la_SOURCES += priv-seccomp.c
 liblldpd_la_CFLAGS  += @SECCOMP_CFLAGS@
 liblldpd_la_LIBADD  += @SECCOMP_LIBS@
+endif
+
+# Add SNMP support if needed
+if USE_SNMP
+noinst_LTLIBRARIES       += liblldpd-snmp.la
+liblldpd_snmp_la_SOURCES  = agent.c agent_priv.c agent.h
+liblldpd_snmp_la_CFLAGS   = $(liblldpd_la_CFLAGS) @NETSNMP_CFLAGS@
+liblldpd_snmp_la_CPPFLAGS = $(liblldpd_la_CPPFLAGS)
+liblldpd_la_LIBADD       += liblldpd-snmp.la
+lldpd_LDADD              += @NETSNMP_LIBS@
 endif
 
 ## Systemtap/DTrace

